service/service.go

```diff
@@ -0,0 +1,312 @@
+            log.Error(err, log.Data{"topic": svc.Topic})
+        }
+        log.Info("error queue consumer will stop when backlog offset reached", log.Data{"backlog_offset": stopAtOffset})
+    }
+
+    var message *sarama.ConsumerMessage
+
+    // We want to stop the processing of the service if consuming from an
+    // error queue if all messages that were initially in the queue have
+    // been cleared using the stopAtOffset
+    running := true
+    for running && (stopAtOffset == -1 || message == nil || message.Offset < stopAtOffset) {
+
+        if message != nil {
+            // Commit the message we've just been processing before starting the next
+            log.Trace("Committing message", log.Data{"offset": message.Offset})
+            svc.Consumer.MarkOffset(message, "")
+            if err := svc.Consumer.CommitOffsets(); err != nil {
+                log.Error(err, log.Data{"offset": message.Offset})
+            }
+        }
+
+        if svc.Retry != nil && svc.Retry.ThrottleRate > 0 {
+            time.Sleep(svc.Retry.ThrottleRate * time.Second)
+        }
+
+        select {
+        case <-c:
+            running = false
+
+        case message = <-svc.Consumer.Messages():
+            // Falls into this block when a message becomes available from consumer
+
+            if message != nil {
+                if message.Offset >= svc.InitialOffset {
+                    log.Info("Received message from Payment Service. Attempting reconciliation...")
+
+                    // GetPayment the payment session first
+                    var pp data.PaymentProcessed
+                    paymentProcessedSchema := &avro.Schema{
+                        Definition: svc.PpSchema,
+                    }
+
+                    err = paymentProcessedSchema.Unmarshal(message.Value, &pp)
+                    if err != nil {
+                        log.Error(err, log.Data{"message_offset": message.Offset})
+                        svc.HandleError(err, message.Offset, &message.Value)
+                        continue
+                    }
+
+                    //Create GetPayment payment session URL
+                    getPaymentURL := svc.PaymentsAPIURL + "/payments/" + pp.ResourceURI
+                    log.Info("Payment URL : " + getPaymentURL)
+                    //Call GetPayment payment session from payments API
+                    paymentResponse,statusCode, err := payment.GetPayment(getPaymentURL, svc.Client, svc.ApiKey)
+                    if err != nil {
+                        log.Error(err, log.Data{"message_offset": message.Offset})
+                        svc.HandleError(err, message.Offset, &paymentResponse)
+                    }
+                    log.Info("Payment Response : ", log.Data{"payment_response": paymentResponse, "status_code":statusCode})
+                    //GetPayment Cost from payment session Cost array
+                    cost, err := paymentResponse.GetCost("cic-report")
```

Comment:
This consumer needs to be generic; we should iterate over all Costs, not just select the `cic-report` costs

Reply:
So it means that if there are multiple costs in a payment data then we will have multiple lines in the eshu file with the same payment reference id but different product code ?

Reply:
I would expect so, yes. Maybe we need to clarify this with an architect though